### PR TITLE
decodeInto() returns number of bytes read

### PR DESCRIPTION
**Pull request by lynn**
Also, fix a critical bug in the definition sorting: array and map types should report their dependencies recursively.

### CODE REVIEW

**Review by skrathbun-rainway — APPROVED** [review]
seems straightforward to me.

File: Laboratory/C++/run_test.sh [hunk 1]
```diff
@@ -1,7 +1,15 @@
...
-time dotnet run --project ../../Compiler --cpp "gen/$1.hpp" --files "../Schemas/$1.bop"
+
+if grep -q Microsoft /proc/version; then
+  # Windows: Visual Studio + WSL to run this script
+  bebopc="../../bin/compiler/Windows-Debug/bebopc.exe"
+else
+  # Linux or Mac
+  bebopc="dotnet run --project ../../Compiler"
+fi
+$bebopc --cpp "gen/$1.hpp" --files "../Schemas/$1.bop"
```
Comment (bugQ) [inline Laboratory/C++/run_test.sh:12]: should this still have a `time` command or no?

Comment (lynn) [reply Laboratory/C++/run_test.sh:12]: I realized the time was not accurate anyway (on Linux/Mac, it includes the startup time of `dotnet` and compilation time, whereas I'm only really interested in the runtime).

Comment (lynn) [reply Laboratory/C++/run_test.sh:12]: Oh shoot, the echo command above still says "Timing" though. Well, I'll fix it next time I run a test :)

Comment (lynn) [reply Laboratory/C++/run_test.sh:12]: Oh shoot, the echo command above still says "Timing" though. Well, I'll fix it next time I run a test :)

**Pull request merged**